src/auth/digest/Config.cc

```diff
@@ -151,85 +156,61 @@ authenticateDigestNonceNew(void)
+    newnonce->cache();
     debugs(29, 5, "created nonce " << newnonce << " at " << newnonce->noncedata.creationtime);
     return newnonce;
 }
 
 static void
 authenticateDigestNonceSetup(void)
 {
-    if (!digest_nonce_cache) {
-        digest_nonce_cache = hash_create((HASHCMP *) strcmp, 7921, hash_string);
-        assert(digest_nonce_cache);
-        eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
-    }
+    eventAdd("Digest nonce cache maintenance", authenticateDigestNonceCacheCleanup, nullptr, static_cast<Auth::Digest::Config*>(Auth::SchemeConfig::Find("digest"))->nonceGCInterval, 1);
 }
 
 void
 authenticateDigestNonceShutdown(void)
 {
-    /*
-     * We empty the cache of any nonces left in there.
-     */
-    digest_nonce_h *nonce;
-
-    if (digest_nonce_cache) {
-        debugs(29, 2, "Shutting down nonce cache");
-        hash_first(digest_nonce_cache);
-
-        while ((nonce = ((digest_nonce_h *) hash_next(digest_nonce_cache)))) {
-            assert(nonce->flags.incache);
-            authDigestNoncePurge(nonce);
-        }
+    // empty the cache of any nonces left in there
+    for (const auto &entry: NonceCache()) {
+        entry.second->purge();
     }
-
     debugs(29, 2, "Nonce cache shutdown");
 }
 
 static void
 authenticateDigestNonceCacheCleanup(void *)
 {
     /*
-     * We walk the hash by noncehex as that is the unique key we
-     * use.  For big hash tables we could consider stepping through
+     * For big hash tables we could consider stepping through
      * the cache, 100/200 entries at a time. Lets see how it flies
      * first.
      */
-    digest_nonce_h *nonce;
     debugs(29, 3, "Cleaning the nonce cache now");
     debugs(29, 3, "Current time: " << current_time.tv_sec);
-    hash_first(digest_nonce_cache);
 
-    while ((nonce = ((digest_nonce_h *) hash_next(digest_nonce_cache)))) {
-        debugs(29, 3, "nonce entry  : " << nonce << " '" << (char *) nonce->key << "'");
+    for (const auto &entry: NonceCache()) {
+        auto nonce = entry.second;
+        if (!nonce)
+            continue;
+
+        debugs(29, 3, "nonce entry  : " << nonce << " '" << nonce->hex() << "'");
         debugs(29, 4, "Creation time: " << nonce->noncedata.creationtime);
 
-        if (nonce->stale()) {
-            debugs(29, 4, "Removing nonce " << (char *) nonce->key << " from cache due to timeout.");
-            assert(nonce->flags.incache);
-            /* invalidate nonce so future requests fail */
-            nonce->flags.valid = false;
+        if (!nonce->validate()) {
+            debugs(29, 4, "Removing nonce " << nonce->hex() << " from cache due to timeout.");
             /* if it is tied to a auth_user, remove the tie */
             if (nonce->user)
                 nonce->user->unlink(nonce);
-            authDigestNoncePurge(nonce);
+            nonce->purge();
```

Comment:
Purging in a range-based for loop is a UB. See another change request about the same problem.

Reply:
BTW, these repeated bugs are related to the same API problem -- an object should not control its own membership in an externally-managed container. Completely solving this problem is not trivial because it requires protecting the cache from invasive external actions, but moving away from nonce::purge() to a Purge(iterator) function may reduce similar problems in future code. I recommend trying to make a few steps in that direction, but I do not insist on those difficult improvements.